ggu/main.c

```diff
@@ -16,11 +16,380 @@
+			rte_be_to_cpu_32(
+			policy->params.u.granted.tx_rate_kb_sec);
+		entry.u.ggu.params.u.granted.cap_expire_sec =
+			rte_be_to_cpu_32(
+			policy->params.u.granted.cap_expire_sec);
+		entry.u.ggu.params.u.granted.renewal_step_ms =
+			rte_be_to_cpu_32(
+			policy->params.u.granted.renewal_step_ms);
+		break;
+
+	case GK_DECLINED:
+		entry.u.ggu.params.u.declined.expire_sec =
+			rte_be_to_cpu_32(policy->params.u.declined.expire_sec);
+		break;
+
+	default:
+		RTE_LOG(ERR, GATEKEEPER, "ggu: impossible policy state %hhu\n",
+			policy->state);
+		return;
+		break;
+	}
+
+	send_gk_cmd(block_id, &entry, ggu_conf->gk);
+}
+
+static void
+process_single_packet(struct rte_mbuf *pkt, struct ggu_config *ggu_conf)
+{
+	uint8_t j;
+	uint8_t v1;
+	uint8_t n1;
+	uint8_t n2;
+	uint8_t n3;
+	uint8_t n4;
+	uint8_t *data;
+	uint8_t *policy_ptr;
+	uint16_t ether_type;
+	struct ether_hdr *eth_hdr;
+	struct ipv4_hdr *ip4hdr;
+	struct ipv6_hdr *ip6hdr;
+	struct udp_hdr *udphdr;
+	uint16_t real_payload_len;
+	uint16_t expected_payload_len;
+	struct ggu_policy policy;
+
+	eth_hdr = rte_pktmbuf_mtod(pkt, struct ether_hdr *);
+	ether_type = rte_be_to_cpu_16(eth_hdr->ether_type);
+
+	switch (ether_type) {
+	case ETHER_TYPE_IPv4:
+		ip4hdr = rte_pktmbuf_mtod_offset(pkt, 
+			struct ipv4_hdr *, sizeof(struct ether_hdr));
+		if (ip4hdr->next_proto_id != IPPROTO_UDP) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received non-UDP packets, IPv4 ntuple filter bug!\n");
+			goto free_packet;
+		}
+
+		if (ip4hdr->dst_addr != ggu_conf->net->back.ip4_addr.s_addr) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received packets not destined to the Gatekeeper server, IPv4 ntuple filter bug!\n");
+			goto free_packet;
+		}
+
+		udphdr = (struct udp_hdr *)&ip4hdr[1];
+		break;
+
+	case ETHER_TYPE_IPv6:
+		ip6hdr = rte_pktmbuf_mtod_offset(pkt, 
+			struct ipv6_hdr *, sizeof(struct ether_hdr));
+		if (ip6hdr->proto != IPPROTO_UDP) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"ggu: received non-UDP packets, IPv6 ntuple filter bug!\n");
+			goto free_packet;
+		}
+
+		/*
+		 * TODO Given that IPv6 ntuple filter doesn't check
+		 * the destination address, it must be done here.
+		 * If the IPv6 packet is not destined to
+		 * the Gatekeeper server, redirect the packet properly.
+		 */
+		if (memcmp(ip6hdr->dst_addr,
+				ggu_conf->net->back.ip6_addr.s6_addr,
+				sizeof(ip6hdr->dst_addr)) != 0) {
+			return;
+		}
+
+		udphdr = (struct udp_hdr *)&ip6hdr[1];
+		break;
+
+	default:
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"ggu: unknown network layer protocol %hu\n",
+			ether_type);
+		goto free_packet;
+		break;
+	}
+
+	/* XXX Check the UDP checksum. */
+
+	if (udphdr->src_port != ggu_conf->ggu_src_port ||
+			udphdr->dst_port != ggu_conf->ggu_dst_port) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"ggu: unknown udp src port %hu, dst port %hu, ntuple filter bug!\n",
+			rte_be_to_cpu_16(udphdr->src_port),
+			rte_be_to_cpu_16(udphdr->dst_port));
+		goto free_packet;
+	}
+
+	data = (uint8_t *)&udphdr[1];
+	v1 = data[0];
```

Comment:
You don't know that `data[0]` exists at this point. What if it's UDP packet with no payload? Malformed packets do arrive in production, either by accident, or malicious behavior. Gatekeeper must not crash due to a malformed packet.

A suggestion to address this problem is to write a function that gets a mbuf and checks if there's the packet has the minimum expected size to do the test. If the packet has size to hold `data[0]', then check the value of 'data[0]' and decide what the next size that should be verified.